src/randomenv.cpp

```diff
@@ -81,6 +99,58 @@ CSHA512& operator<<(CSHA512& hasher, const T& data) {
     return hasher;
 }
 
+#ifndef WIN32
+void AddSockaddr(CSHA512& hasher, const struct sockaddr *addr)
+{
+    if (addr == nullptr) return;
+    switch (addr->sa_family) {
+    case AF_INET:
+        hasher.Write((const unsigned char*)addr, sizeof(sockaddr_in));
+        break;
+    case AF_INET6:
+        hasher.Write((const unsigned char*)addr, sizeof(sockaddr_in6));
+        break;
+    default:
+        hasher.Write((const unsigned char*)&addr->sa_family, sizeof(addr->sa_family));
+    }
+}
+
+void AddStat(CSHA512& hasher, const struct stat& sb)
+{
+    hasher << sb.st_dev << sb.st_ino << sb.st_mode << sb.st_nlink << sb.st_uid << sb.st_gid << sb.st_size << sb.st_blksize << sb.st_blocks << sb.st_atime << sb.st_mtime << sb.st_ctime;
+}
+
+void AddFile(CSHA512& hasher, const char *path)
+{
+    struct stat sb;
+    int f = open(path, O_RDONLY);
+    if (f != -1) {
+        unsigned char fbuf[1024];
+        int n;
+        hasher.Write((const unsigned char*)&f, sizeof(f));
+        if (fstat(f, &sb) == 0) {
+            AddStat(hasher, sb);
+        }
+        do {
+            n = read(f, fbuf, sizeof(fbuf));
+            if (n > 0) hasher.Write(fbuf, n);
+            /* not bothering with EINTR handling. */
+        } while (n == sizeof(fbuf));
```

Comment:
Might you consider only reading up to a certain number of bytes? I'm not sure you're guaranteed that some of these files won't be enormous.

Reply:
Done.